lib/rules/prefer-const.js

```diff
@@ -361,109 +343,57 @@ module.exports = {
-                        fix: shouldFix ? fixer => fixer.replaceText(sourceCode.getFirstToken(varDeclParent), "const") : null
-                    });
-                });
-            }
-        }
 
         return {
-            "Program:exit"() {
-                groupByDestructuring(variables, ignoreReadBeforeAssign).forEach(checkGroup);
-            },
-
             VariableDeclaration(node) {
                 if (node.kind === "let" && !isInitOfForStatement(node)) {
-                    variables.push(...context.getDeclaredVariables(node));
+                    const nodesToReport = [];
+                    let shouldFix = true;
+
+                    groupByDestructuring(context.getDeclaredVariables(node), ignoreReadBeforeAssign)
+                        .forEach(identifiers => {
+
+                            /*
+                             * 'identifiers' is an array of Identifier nodes from the same group.
+                             * In simple declaration or assignment cases, the length of the array is 1.
+                             * In destructuring cases, the length of the array can be 2 or more.
+                             *
+                             * Each element in this array is the result of 'getIdentifierIfShouldBeConst()',
+                             * so it will be null if the corresponding identifier cannot be const.
+                             * Elements that are not null are the ones that could be const.
+                             */
+                            const couldBeConstIdentifiers = identifiers.filter(Boolean);
+
+                            /*
+                             * If all elements from the same group could be const, report them.
+                             * Otherwise (destructuring where only some could be const), report depending on the configuration.
+                             */
+                            if (couldBeConstIdentifiers.length &&
+                                (couldBeConstIdentifiers.length === identifiers.length || shouldMatchAnyDestructuredVariable)) {
+                                nodesToReport.push(...couldBeConstIdentifiers);
+                            }
+
+                            // If at least one identifier cannot be const, the whole declaration cannot be auto-fixed to const
+                            if (couldBeConstIdentifiers.length < identifiers.length) {
+                                shouldFix = false;
+                            }
+                        });
+
+                    // Don't do a fix unless all variables are initialized (or it's in a for-in or for-of loop)
+                    shouldFix = shouldFix &&
+                        (node.parent.type === "ForInStatement" || node.parent.type === "ForOfStatement" ||
+                            node.declarations.every(declaration => declaration.init));
+
+                    const declarationKindToken = sourceCode.getFirstToken(node);
```

Comment:
It would probably be better to use a predicate here, in case of Typescript `declare let` or other potential token prefixes:

```suggestion
                    const declarationKindToken = sourceCode.getFirstToken(node, t => t.value === node.kind);
```

Reply:
I'll fix this, it was inherited from `checkGroup`, as well as the autofix for each variable which isn't necessary now when we have the whole declaration.

Speaking of TypeScript, it seems that `@typescript-eslint/parser` allows `let let = 1`, so one fix without the other might produce changing several tokens.

Reply:
It's fixed now within the second commit (multiple autofixes are also removed).

While it is most likely good to support potential prefixes, this particular example shouldn't produce a problem (like [#11441](https://github.com/eslint/eslint/issues/11441)), because only initialized variables are auto-fixed and the following is not a valid TypeScript:

```ts
declare let x = 1; // tsc error `Initializers are not allowed in ambient contexts`
```

Nevertheless, `@typescript-eslint/parser` treats this as a valid code, so the autofix without this correction would indeed replace `declare`.